futures-util/src/future/abortable.rs

```diff
@@ -0,0 +1,131 @@
+        Abortable {
+            future: future,
+            inner: reg.inner,
+        }
+    }
+
+    unsafe_pinned!(future: T);
+}
+
+/// A registration handle for a `Abortable` future.
+/// Values of this type can be acquired from `AbortHandle::new` and are used
+/// in calls to `Abortable::new`.
+#[derive(Debug, Clone)]
+pub struct AbortRegistration {
+    inner: Arc<AbortInner>,
+}
+
+/// A handle to a `Abortable` future.
+#[derive(Debug, Clone)]
+pub struct AbortHandle {
+    inner: Arc<AbortInner>,
+}
+
+impl AbortHandle {
+    /// Creates a new `AbortHandle` which can be used to abort a running future.
+    /// This function also returns a `AbortRegistration` which is used to
+    /// pair the handle with a `Abortable` future.
+    ///
+    /// This function is usually paired with a call to `Abortable::new`.
+    pub fn new() -> (Self, AbortRegistration) {
+        let inner = Arc::new(AbortInner {
+            waker: AtomicWaker::new(),
+            cancel: AtomicBool::new(false),
+        });
+
+        (
+            AbortHandle {
+                inner: inner.clone(),
+            },
+            AbortRegistration {
+                inner: inner,
+            },
+        )
+    }
+}
+
+// Inner type storing the waker to awaken and a bool indicating that it
+// should be cancelled.
+#[derive(Debug)]
+struct AbortInner {
+    waker: AtomicWaker,
+    cancel: AtomicBool,
+}
+
+/// Creates a new `Abortable` future and a `AbortHandle` which can be used to stop it.
+///
+/// This function is a convenient (but less flexible) alternative to calling
+/// `AbortHandle::new` and `Abortable::new` manually.
+pub fn abortable<T>(future: T) -> (Abortable<T>, AbortHandle)
+    where T: Future
+{
+    let (handle, reg) = AbortHandle::new();
+    (
+        Abortable::new(future, reg),
+        handle,
+    )
+}
+
+/// Indicator that the `Abortable` future was aborted.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct Aborted;
+
+impl<T> Future for Abortable<T> where T: Future {
+    type Output = Result<T::Output, Aborted>;
+
+    fn poll(mut self: PinMut<Self>, cx: &mut task::Context) -> Poll<Self::Output> {
+        // Check if the future has been aborted
+        if self.inner.cancel.load(Ordering::Relaxed) {
+            return Poll::Ready(Err(Aborted))
+        }
+
+        // attempt to complete the future
+        if let Poll::Ready(x) = self.future().poll(cx) {
+            return Poll::Ready(Ok(x))
+        }
+
+        // Register to receive a wakeup if the future is aborted in the... future
+        self.inner.waker.register(cx.waker());
+
+        // Check to see if the future was aborted between the first check and
+        // registration
+        if self.inner.cancel.load(Ordering::Relaxed) {
+            return Poll::Ready(Err(Aborted))
+        }
+
+        Poll::Pending
+    }
+}
+
+impl AbortHandle {
+    /// Abort the `Abortable` future associated with this handle.
```

Comment:
The docs should mention that `abort()` will abort the future ASAP, but does not block until it has actually been done.

Reply:
Aborting it isn't an operation that requires a delay, so I don't think it makes sense to refer to it as blocking/non-blocking.

Reply:
Here's what I'd like the docs to say:
Notifies the `Abortable` future associated with this handle that it should abort. Note that if the future is currently being polled on another thread, it will not immediately stop running, instead it will continue to run until its `poll` method returns.

Reply:
Done.